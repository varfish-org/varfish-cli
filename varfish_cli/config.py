"""Common configuration code."""

import os

try:
    import tomllib
    from tomllib import TOMLDecodeError
except ImportError:
    import toml as tomllib
    from toml import TomlDecodeError as TOMLDecodeError

import typing

from logzero import logger
import pydantic
import typer


class CommonOptions(pydantic.BaseModel):
    """Shared options for all subcommands."""

    #: Whether verbose output is enabled.
    verbose: bool = False
    #: Whether to verify SSL.
    verify_ssl: bool = True
    #: Path to configuration file.
    config: typing.Optional[str] = None
    #: VarFish server URL key to use.
    varfish_server_url: typing.Optional[str] = None
    #: VarFish API token to use.
    varfish_api_token: typing.Optional[pydantic.SecretStr] = None


def load_config(config_path: str) -> typing.Tuple[typing.Optional[str], typing.Optional[str]]:
    """Load configuration and return server URL and API token.

    :returns: Tuple of ``(server_url, api_token)``
    """
    config_path = os.path.expanduser(os.path.expandvars(config_path))

    toml_varfish_server_url: typing.Optional[str] = None
    toml_varfish_api_token: typing.Optional[str] = None

    if not os.path.exists(config_path):
        logger.info("could not find configuration file %s.", config_path)
    else:
        logger.debug("loading configuration from %s", config_path)
        with open(config_path, "rt") as tomlf:
            try:
                config_toml = tomllib.loads(tomlf.read())
            except TOMLDecodeError as e:
                logger.error("could not parse configuration file %s: %s", config_path, e)
                raise typer.Exit(1)
            toml_varfish_server_url = config_toml.get("global", {}).get("varfish_server_url")
            if toml_varfish_server_url:
                logger.debug("using global/varfish_server_url from %s", config_path)
            else:
                logger.debug("global/varfish_server_url not set in %s", config_path)
            toml_varfish_api_token = config_toml.get("global", {}).get("varfish_api_token")
            if toml_varfish_api_token:
                logger.debug("using global/varfish_server_url from %s", config_path)
            else:
                logger.debug("global/varfish_api_token not set in %s", config_path)

    return toml_varfish_server_url, toml_varfish_api_token
