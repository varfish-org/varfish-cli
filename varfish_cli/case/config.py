"""Configuration classes for ``varfish-cli case *`` commands."""

import json
import typing
import uuid

import attr

from varfish_cli.api import CONVERTER, models
from varfish_cli.common import CommonConfig, OutputFormat


@attr.s(frozen=True, auto_attribs=True)
class CaseConfig:
    """Configuration for the ``varfish-cli case`` command."""

    #: Global configuration
    global_config: CommonConfig

    #: Path to output file
    output_file: str = "-"

    #: Output format
    output_format: OutputFormat = OutputFormat.TABLE

    #: delimiter for CSV output
    output_delimiter: str = ","

    #: Fields to use for output.
    output_fields: typing.Optional[typing.List[str]] = []

    @staticmethod
    def create(args, global_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseConfig(
            output_format=OutputFormat(args.output_format),
            output_file=args.output_file,
            output_delimiter=args.output_delimiter,
            output_fields=args.output_fields.split(",") if args.output_fields else [],
            global_config=global_config,
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseListConfig:
    """Configuration for the ``varfish-cli case list`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the case to pull.
    project_uuid: uuid.UUID

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseListConfig(case_config=case_config, project_uuid=args.project_uuid)


@attr.s(frozen=True, auto_attribs=True)
class CaseListImportInfoConfig:
    """Configuration for the ``varfish-cli case list-import-info`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the case to pull.
    project_uuid: uuid.UUID

    #: Optionally, owner to query for.
    owner: typing.Optional[str] = None

    @staticmethod
    def create(args, case_config, toml_config=None):
        # toml_config = toml_config or {}
        return CaseListImportInfoConfig(
            case_config=case_config, project_uuid=args.project_uuid, owner=args.owner
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseCreateImportInfoConfig:
    """Configuration for the ``varfish-cli case create-import-info`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: Suffix to append to the case name.
    case_name_suffix: str

    #: UUID of the case to pull.
    project_uuid: uuid.UUID

    #: Path to files to import.
    paths: typing.List[str]

    #: Regular expression to use for modifying family.
    strip_family_regex: str

    #: Whether to force resubmittal of old
    resubmit: bool

    #: Whether to force creation of fresh case import info.
    force_fresh: bool

    #: Expected genome build.
    genomebuild: str

    @staticmethod
    def create(args, case_config, strip_family_regex, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseCreateImportInfoConfig(
            case_config=case_config,
            project_uuid=args.project_uuid,
            paths=args.paths,
            strip_family_regex=args.strip_family_regex,
            case_name_suffix=args.case_name_suffix,
            resubmit=args.resubmit,
            force_fresh=args.force_fresh,
            genomebuild=args.genomebuild,
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseSmallVariantQueryListConfig:
    """Configuration for the ``varfish-cli case small-var-query-list`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the case to list for.
    case_uuid: uuid.UUID

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseSmallVariantQueryListConfig(case_config=case_config, case_uuid=args.case_uuid)


@attr.s(frozen=True, auto_attribs=True)
class CaseSmallVariantQueryCreateConfig:
    """Configuration for the ``varfish-cli case small-var-query-create`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the small variant query to retrieve.
    case_uuid: uuid.UUID

    #: The query settings to set.
    query_settings: models.CaseQuerySettingsV1

    #: The name of the query to set, if any.
    name: typing.Optional[str] = None

    #: Whether or not to make the query public, if any.
    public: bool = False

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        if args.query_settings.startswith("@"):
            with open(args.query_settings[1:], "rt") as inputf:
                query_settings = json.load(inputf)
        else:
            query_settings = json.loads(args.args.query_settings.startswith)
        # Allow both reading raw query_settings and response from query settings shortcuts
        if "query_settings" in query_settings:
            query_settings = query_settings["query_settings"]
        return CaseSmallVariantQueryCreateConfig(
            case_config=case_config,
            case_uuid=args.case_uuid,
            query_settings=CONVERTER.structure(query_settings, models.CaseQuerySettingsV1),
            name=args.name,
            public=args.public,
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseSmallVariantQueryShortcut:
    """Configuration for the ``varfish-cli case small-var-query-shortcut`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the small variant query to retrieve.
    case_uuid: uuid.UUID

    #: The database to use
    database: str

    #: The quick preset to use
    quick_preset: str

    #: Setting in category inheritance
    inheritance: typing.Optional[str] = None

    #: Setting in category frequency
    frequency: typing.Optional[str] = None

    #: Setting in category impact
    impact: typing.Optional[str] = None

    #: Setting in category quality
    quality: typing.Optional[str] = None

    #: Setting in category chromosomes
    chromosomes: typing.Optional[str] = None

    #: Setting in category flags_etc
    flags_etc: typing.Optional[str] = None

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseSmallVariantQueryShortcut(
            case_config=case_config,
            case_uuid=args.case_uuid,
            database=args.database,
            quick_preset=args.quick_preset,
            inheritance=args.inheritance,
            frequency=args.frequency,
            impact=args.impact,
            quality=args.quality,
            chromosomes=args.chromosomes,
            flags_etc=args.flags_etc,
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseSmallVariantQueryRetrieveConfig:
    """Configuration for the ``varfish-cli case small-var-query-retrieve`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the small variant query to retrieve.
    query_uuid: uuid.UUID

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseSmallVariantQueryRetrieveConfig(
            case_config=case_config, query_uuid=args.query_uuid
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseSmallVariantQueryStatusConfig:
    """Configuration for the ``varfish-cli case small-var-query-status`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the small variant query to retrieve.
    query_uuid: uuid.UUID

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseSmallVariantQueryStatusConfig(
            case_config=case_config, query_uuid=args.query_uuid
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseSmallVariantQueryUpdateConfig:
    """Configuration for the ``varfish-cli case small-var-query-update`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the small variant query to retrieve.
    query_uuid: uuid.UUID

    #: The name of the query to set, if any.
    name: typing.Optional[str] = None

    #: Whether or not to make the query public, if any.
    public: typing.Optional[bool] = None

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseSmallVariantQueryUpdateConfig(
            case_config=case_config, query_uuid=args.query_uuid, name=args.name, public=args.public
        )


@attr.s(frozen=True, auto_attribs=True)
class CaseSmallVariantQueryFetchResultsConfig:
    """Configuration for the ``varfish-cli case small-var-query-fetch-results`` command."""

    #: Case configuration.
    case_config: CaseConfig

    #: UUID of the small variant query to retrieve.
    query_uuid: uuid.UUID

    @staticmethod
    def create(args, case_config, toml_config=None):
        _ = toml_config
        # toml_config = toml_config or {}
        return CaseSmallVariantQueryFetchResultsConfig(
            case_config=case_config, query_uuid=args.query_uuid
        )
