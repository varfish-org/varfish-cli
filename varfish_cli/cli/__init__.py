"""Code for main entry point."""

import logging
import typing

import logzero
import typer

from varfish_cli import __version__
from varfish_cli.cli import cases, importer, projects, tools, varannos
from varfish_cli.config import CommonOptions, load_config
from varfish_cli.exceptions import InvalidConfiguration

#: Paths to search the global configuration in.
DEFAULT_PATH_VARFISHRC = "~/.varfishrc.toml"


def version_callback(value: bool):
    """Callback when called with 'version' or '--version'"""
    if value:
        print(f"varfish-cli {__version__}")
        raise typer.Exit()


#: Main CLI ``Typer`` object.
app = typer.Typer(no_args_is_help=True, pretty_exceptions_enable=False)

# Register all sub commands.
app.add_typer(varannos.app, name="varannos", help="Subcommands for 'varannos' API")
app.add_typer(projects.app, name="projects", help="Subcommands for 'project' API")
app.add_typer(cases.app, name="cases", help="Subcommands for 'cases' API")
app.add_typer(importer.app, name="importer", help="Subcommands for 'importer' API")
app.add_typer(tools.app, name="tools", help="Subcommands for 'tools' API")


@app.command("version")
def main_version():
    """Print version as "varfish-cli $version"."""
    version_callback(True)


@app.callback()
def main(
    ctx: typer.Context,
    version: typing.Annotated[
        typing.Optional[bool], typer.Option("--version", callback=version_callback)
    ] = None,
    verbose: typing.Annotated[
        bool, typer.Option("--verbose", "-v", help="Enable verbose output")
    ] = False,
    verify_ssl: typing.Annotated[
        bool, typer.Option("--verify-ssl/--no-verify-ssl", help="Disable SSL verification")
    ] = True,
    config_path: typing.Annotated[
        str,
        typer.Option("--config-path", help="Path to configuration file", envvar="VARFISH_RC_PATH"),
    ] = DEFAULT_PATH_VARFISHRC,
    varfish_server_url: typing.Annotated[
        typing.Optional[str],
        typer.Option(
            "--varfish-server-url",
            help=(
                "VarFish server URL key to use, defaults to env VARFISH_SERVER_URL or read "
                "from configfile"
            ),
            envvar="VARFISH_SERVER_URL",
        ),
    ] = None,
    varfish_api_token: typing.Annotated[
        typing.Optional[str],
        typer.Option(
            "--varfish-server-url",
            help=(
                "VarFish API token to use, defaults to env VARFISH_API_TOKEN or read from "
                "configfile"
            ),
            envvar="VARFISH_API_TOKEN",
        ),
    ] = None,
):
    """Callback for main entry point

    This function handles the global configuration from configuration file,
    environment variables, and command line (in increasing priority).
    """
    _ = version
    # Setup logging
    if verbose:  # pragma: no cover
        level = logging.DEBUG
    else:
        # Remove module name and line number if not running in debug mode.s
        formatter = logzero.LogFormatter(
            fmt="%(color)s[%(levelname)1.1s %(asctime)s]%(end_color)s %(message)s"
        )
        logzero.formatter(formatter)
        level = logging.INFO
    logzero.loglevel(level=level)

    # Load configuration file
    toml_varfish_server_url, toml_varfish_api_token = load_config(config_path)
    if toml_varfish_server_url and not varfish_server_url:
        varfish_server_url = toml_varfish_server_url
    if toml_varfish_api_token and not varfish_api_token:
        varfish_api_token = toml_varfish_api_token

    if not varfish_server_url or not varfish_api_token:
        raise InvalidConfiguration("Need to specify server URL and API token")

    # Construct common options
    ctx.obj = CommonOptions(
        verbose=verbose,
        verify_ssl=verify_ssl,
        config_path=config_path,
        varfish_server_url=varfish_server_url,
        varfish_api_token=varfish_api_token,
    )


#: Define ``typer`` object that can be called from ``__main__.py``.
typer_click_object = typer.main.get_command(app)
