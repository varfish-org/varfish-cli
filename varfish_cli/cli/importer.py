"""Implementation of varfish-cli subcommand "importer *"."""

import sys
import typing
import uuid

from logzero import logger
import typer

from varfish_cli import api, common

#: The ``Typer`` instance to use for the ``cases`` sub command.
app = typer.Typer(no_args_is_help=True)


@app.command("caseimportinfo-list")
def run(
    ctx: typer.Context,
    project_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of project to list cases for")
    ],
    owner: typing.Annotated[
        typing.Optional[str],
        typer.Option("--owner", help="Optionally, name of owner to filter for"),
    ] = None,
    output_file: typing.Annotated[
        str, typer.Option("--output-file", help="Path to file to write to")
    ] = "-",
):
    # config, toml_config, args, _parser, _subparser, file=sys.stdout):
    """List case import infos."""
    common_options: common.CommonConfig = ctx.obj
    logger.info("Listing CaseImportInfo records")
    if owner:
        logger.info("- filter owner: %s" % owner)
    else:
        logger.info("- filter owner: <any>")

    res = api.case_import_info_list(
        server_url=common_options.varfish_server_url,
        api_token=common_options.varfish_api_token.get_secret_value(),
        project_uuid=project_uuid,
        owner=owner,
        verify_ssl=common_options.verify_ssl,
    )

    def do_print(file: typing.TextIO):
        print("Case Import Info List", file=file)
        print("=====================", file=file)
        print(file=file)
        if not res:
            print("No records found.", file=file)
        for info in res:
            print("- uuid: %s" % repr(str(info.sodar_uuid)), file=file)
            print("  release: %s" % repr(info.release.value), file=file)
            print("  owner: %s" % repr(info.owner), file=file)
            print("  name: %s" % repr(info.name), file=file)
            print("  index: %s" % repr(info.index), file=file)
            if info.pedigree:
                print("  members:", file=file)
                for member in info.pedigree:
                    print(
                        "    - { name: %s, father: %s, mother: %s, sex: %s, affected: %s }"
                        % tuple(
                            [
                                repr(x)
                                for x in (
                                    member.name,
                                    member.father,
                                    member.mother,
                                    member.sex,
                                    member.affected,
                                )
                            ]
                        ),
                        file=file,
                    )
            else:
                print("  members: []", file=file)
            print(file=file)
        file.flush()

    logger.info("Writing output")
    if output_file == "-":
        do_print(sys.stdout)
    else:
        with open(output_file, "wt") as outputf:
            do_print(outputf)

    logger.info("All done. Have a nice day!")
