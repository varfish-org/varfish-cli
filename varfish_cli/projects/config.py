import typing

import attr

from varfish_cli.common import CommonConfig, OutputFormat


@attr.s(frozen=True, auto_attribs=True)
class ProjectsConfig:
    """Configuration for the ``varfish-cli project`` command."""

    #: Global configuration
    global_config: CommonConfig

    #: Path to output file
    output_file: str = "-"

    #: Output format
    output_format: OutputFormat = OutputFormat.TABLE

    #: delimiter for CSV output
    output_delimiter: str = ","

    #: Fields to use for output.
    output_fields: typing.Optional[typing.List[str]] = []

    @staticmethod
    def create(args, global_config, toml_config=None):
        _ = toml_config
        return ProjectsConfig(
            output_format=OutputFormat(args.output_format),
            output_file=args.output_file,
            output_delimiter=args.output_delimiter,
            output_fields=args.output_fields.split(",") if args.output_fields else [],
            global_config=global_config,
        )


@attr.s(frozen=True, auto_attribs=True)
class ProjectsListConfig:
    """Configuration for the ``varfish-cli projects project-list`` command."""

    #: Project configuration
    projects_config: ProjectsConfig

    @staticmethod
    def create(args, projects_config, toml_config=None):
        _ = args
        _ = toml_config
        return ProjectsListConfig(
            projects_config=projects_config,
        )
