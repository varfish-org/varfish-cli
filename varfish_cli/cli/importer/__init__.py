"""Implementation of varfish-cli subcommand "importer *"."""

import sys
import typing
import uuid

from logzero import logger
import typer

from varfish_cli import api
from varfish_cli.api import GenomeBuild
from varfish_cli.cli.importer.create import CaseImporter, CaseImportOptions
from varfish_cli.config import CommonOptions

#: The ``Typer`` instance to use for the ``importer`` sub command.
app = typer.Typer(no_args_is_help=True)


@app.command("caseimportinfo-list")
def run(
    ctx: typer.Context,
    project_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of project to list cases for")
    ],
    owner: typing.Annotated[
        typing.Optional[str],
        typer.Option("--owner", help="Optionally, name of owner to filter for"),
    ] = None,
    output_file: typing.Annotated[
        str, typer.Option("--output-file", help="Path to file to write to")
    ] = "-",
):
    """List case import infos."""
    common_options: CommonOptions = ctx.obj
    logger.info("Listing CaseImportInfo records")
    if owner:
        logger.info("- filter owner: %s" % owner)
    else:
        logger.info("- filter owner: <any>")

    res = api.case_import_info_list(
        server_url=common_options.varfish_server_url,
        api_token=common_options.varfish_api_token.get_secret_value(),
        project_uuid=project_uuid,
        owner=owner,
        verify_ssl=common_options.verify_ssl,
    )

    def do_print(file: typing.TextIO):
        print("Case Import Info List", file=file)
        print("=====================", file=file)
        print(file=file)
        if not res:
            print("No records found.", file=file)
        for info in res:
            print("- uuid: %s" % repr(str(info.sodar_uuid)), file=file)
            print("  release: %s" % repr(info.release.value), file=file)
            print("  owner: %s" % repr(info.owner), file=file)
            print("  name: %s" % repr(info.name), file=file)
            print("  index: %s" % repr(info.index), file=file)
            if info.pedigree:
                print("  members:", file=file)
                for member in info.pedigree:
                    print(
                        "    - { name: %s, father: %s, mother: %s, sex: %s, affected: %s }"
                        % tuple(
                            [
                                repr(x)
                                for x in (
                                    member.name,
                                    member.father,
                                    member.mother,
                                    member.sex,
                                    member.affected,
                                )
                            ]
                        ),
                        file=file,
                    )
            else:
                print("  members: []", file=file)
            print(file=file)
        file.flush()

    logger.info("Writing output")
    if output_file == "-":
        do_print(sys.stdout)
    else:
        with open(output_file, "wt") as outputf:
            do_print(outputf)

    logger.info("All done. Have a nice day!")


@app.command("caseimportinfo-create")
def cli_caseimportinfo_create(
    ctx: typer.Context,
    project_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of project to list cases for")
    ],
    paths: typing.Annotated[
        typing.List[str],
        typer.Argument(
            ...,
            help="Path(s) to files to use for the import. Must include PED, and annotation TSV files",
        ),
    ] = None,
    strip_family_regex: typing.Annotated[
        str,
        typer.Option("--strip-family-regex", help="Regular expression to process family name with"),
    ] = "^FAM_",
    case_name_suffix: typing.Annotated[
        str, typer.Option("--case-name-suffix", help="Suffix to append to case name")
    ] = "",
    force_fresh: typing.Annotated[
        bool,
        typer.Option(
            "--force-fresh/--no-force-fresh",
            help="Force using fresh case import even if old draft found",
        ),
    ] = False,
    resubmit: typing.Annotated[
        bool,
        typer.Option(
            "--resubmit/--no-resubmit",
            help="Force resubmission of cases in submit state",
        ),
    ] = False,
    genomebuild: typing.Annotated[
        GenomeBuild,
        typer.Option(
            "--genomebuild",
            help="The genome build (GRCh37/GRCh38) of this case, defaults to GRCh37.",
        ),
    ] = GenomeBuild.GRCH37.value,
):
    logger.info("Creating CaseImportInfo object...")
    common_options: CommonOptions = ctx.obj
    case_importer = CaseImporter(
        options=CaseImportOptions(
            paths=paths,
            genomebuild=genomebuild,
            strip_family_regex=strip_family_regex,
            case_name_suffix=case_name_suffix,
            force_fresh=force_fresh,
            resubmit=resubmit,
            project_uuid=project_uuid,
        ),
        common_options=common_options,
    )
    case_importer.run()
    logger.info("All done. Have a nice day!")
