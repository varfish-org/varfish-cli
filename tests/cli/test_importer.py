"""Test CLI for importer API."""

import glob
import json
import types
import typing
import uuid

import pytest
from pytest_mock import MockerFixture
from requests_mock import ANY
from requests_mock.mocker import Mocker as RequestsMocker
from syrupy import SnapshotAssertion
from typer.testing import CliRunner

from tests.conftest import FakeFs
from varfish_cli.cli import app
from varfish_cli.cli.importer import cli_caseimportinfo_create
from varfish_cli.config import CommonOptions


@pytest.fixture
def caseimportinfo_list_result_empty() -> typing.List[typing.Any]:
    return []


def test_varannoset_list_empty(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    caseimportinfo_list_result_empty: typing.List[typing.Any],
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    project_uuid = str(uuid.uuid4())
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/importer/api/case-import-info/{project_uuid}/",
        json=caseimportinfo_list_result_empty,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(app, ["--verbose", "importer", "caseimportinfo-list", project_uuid])

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == "Case Import Info List\n=====================\n\nNo records found.\n"


@pytest.fixture
def caseimportinfo_list_result_two_elements() -> typing.List[typing.Any]:
    with open("tests/cli/data/caseimportinfo-list.len-2.json", "rt") as inputf:
        return json.load(inputf)


def test_varannoset_list_one_element(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    caseimportinfo_list_result_two_elements: typing.List[typing.Any],
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    project_uuid = str(uuid.uuid4())
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/importer/api/case-import-info/{project_uuid}/",
        json=caseimportinfo_list_result_two_elements,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(app, ["--verbose", "importer", "caseimportinfo-list", project_uuid])

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_caseimportinfo_create_raw_func_call(
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    caseimportinfo_list_result_two_elements: typing.List[typing.Any],
    mocker: MockerFixture,
):
    """Test through function call and not runner.

    We create a stubbed out replay of the requests communication and only check
    results against snapshots. This is not much more than a smoke test or a
    change test.
    """
    fake_fs_configured.fs.add_real_directory("tests/data/importer")

    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    project_uuid = "5b1d876d-5ea9-426f-9f8f-f82b18830e16"
    host, token = fake_conn
    m_any = requests_mock.register_uri(
        ANY, ANY, request_headers={"Authorization": f"Token {token}"}, json=[]
    )
    m_case_import_info = requests_mock.post(
        f"{host}/importer/api/case-import-info/{project_uuid}/",
        request_headers={"Authorization": f"Token {token}"},
        json=caseimportinfo_list_result_two_elements[0],
    )
    m_case_bam_qc_file = requests_mock.post(
        f"{host}/importer/api/bam-qc-file/8adbf84e-adb2-4c5d-9b5a-2a8194307b79/",
        request_headers={"Authorization": f"Token {token}"},
        json=caseimportinfo_list_result_two_elements[0]["bam_qc_files"][0],
    )
    m_case_variant_set_import_info_create = requests_mock.post(
        f"{host}/importer/api/variant-set-import-info/8adbf84e-adb2-4c5d-9b5a-2a8194307b79/",
        request_headers={"Authorization": f"Token {token}"},
        json=caseimportinfo_list_result_two_elements[0]["variant_sets"][0],
    )
    m_case_genotype_file = requests_mock.post(
        f"{host}/importer/api/genotype-file/104f900b-fa9c-4bd5-a119-35544293fe7f/",
        request_headers={"Authorization": f"Token {token}"},
        json=caseimportinfo_list_result_two_elements[0]["variant_sets"][0]["genotype_files"][0],
    )
    m_case_db_info_file = requests_mock.post(
        f"{host}/importer/api/database-info-file/104f900b-fa9c-4bd5-a119-35544293fe7f/",
        request_headers={"Authorization": f"Token {token}"},
        json=caseimportinfo_list_result_two_elements[0]["variant_sets"][0]["db_info_files"][0],
    )
    m_case_variant_set_import_info_update = requests_mock.put(
        (
            f"{host}/importer/api/variant-set-import-info/8adbf84e-adb2-4c5d-9b5a-2a8194307b79/"
            "104f900b-fa9c-4bd5-a119-35544293fe7f/"
        ),
        request_headers={"Authorization": f"Token {token}"},
        json=caseimportinfo_list_result_two_elements[0]["variant_sets"][0],
    )
    m_case_import_info_update = requests_mock.put(
        (
            f"{host}/importer/api/case-import-info/5b1d876d-5ea9-426f-9f8f-f82b18830e16/"
            "8adbf84e-adb2-4c5d-9b5a-2a8194307b79/"
        ),
        request_headers={"Authorization": f"Token {token}"},
        json=caseimportinfo_list_result_two_elements[0],
    )
    ctx = types.SimpleNamespace(
        obj=CommonOptions(
            verbose=True,
            verify_ssl=False,
            config=None,
            varfish_server_url=host,
            varfish_api_token=token,
        )
    )
    cli_caseimportinfo_create(
        ctx=ctx,
        project_uuid=project_uuid,
        paths=list(sorted(glob.glob("tests/data/importer/*"))),
    )

    assert m_any.call_count == 10
    assert m_case_import_info.call_count == 1
    assert m_case_bam_qc_file.call_count == 1
    assert m_case_variant_set_import_info_create.call_count == 2
    assert m_case_genotype_file.call_count == 2
    assert m_case_db_info_file.call_count == 2
    assert m_case_variant_set_import_info_update.call_count == 2
    assert m_case_import_info_update.call_count == 1

    mocker.stopall()
