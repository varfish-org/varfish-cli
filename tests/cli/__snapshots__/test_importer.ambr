# serializer version: 1
# name: test_varannoset_list_one_element
  '''
  Case Import Info List
  =====================
  
  - uuid: '8adbf84e-adb2-4c5d-9b5a-2a8194307b79'
    release: 'GRCh37'
    owner: 'root'
    name: 'NA12878'
    index: 'NA12878-N1-DNA1-WES1'
    members:
      - { name: 'NA12878-N1-DNA1-WES1', father: 'NA12891-N1-DNA1-WES1', mother: 'NA12892-N1-DNA1-WES1', sex: 2, affected: 2 }
      - { name: 'NA12891-N1-DNA1-WES1', father: '0', mother: '0', sex: 1, affected: 1 }
      - { name: 'NA12892-N1-DNA1-WES1', father: '0', mother: '0', sex: 2, affected: 1 }
  
  - uuid: '539fca3a-554f-4595-bfb3-b692f8ddef9e'
    release: 'GRCh37'
    owner: 'root'
    name: 'Case_3_index'
    index: 'Case_3_index-N1-DNA1-WGS1'
    members:
      - { name: 'Case_3_index-N1-DNA1-WGS1', father: 'Case_3_father-N1-DNA1-WGS1', mother: 'Case_3_mother-N1-DNA1-WGS1', sex: 1, affected: 2 }
      - { name: 'Case_3_father-N1-DNA1-WGS1', father: '0', mother: '0', sex: 1, affected: 1 }
      - { name: 'Case_3_mother-N1-DNA1-WGS1', father: '0', mother: '0', sex: 2, affected: 1 }
  
  
  '''
# ---
